python/pyspark/sql/streaming/listener.py

```diff
@@ -497,6 +499,19 @@ def fromJson(cls, j: Dict[str, Any]) -> "StreamingQueryProgress":
             else {},
         )
 
+    def __getitem__(self, key):
+        # Before Spark 4.0, StreamingQuery.lastProgress returns a dict, which casts id and runId
+        # to string. To prevent breaking change, also cast them to string when accessed with
+        # __getitem__.
+        if key == "id" or key == "runId":
```

Comment:
I'm not sure if this is really needed

Reply:
Because there would be lots of breaking changes (e.g. now the `sources` method also return the actual `SourceProgress`
https://github.com/apache/spark/blob/ef7a116828d1781dde35d422550ff369c548a6aa/python/pyspark/sql/streaming/listener.py#L595

let me also make these subclass of dict...